Make past_tallies a sparse table

cpeel opened the pull request:
# Background
_Background pulled from #1190_

The `past_tallies` table stores historical round, user, and team statistics (`current_tallies` stores the current day's statistics) including the number of pages done that day (`tally_delta`). Once a day the site takes a snapshot of `current_tallies`, calculates the delta since the last snapshot, and adds a row in the `past_tallies` table for every `[tally_type, tally_name, holder_id]` tuple.

After a user has done a page in a round, they have an entry for that round in `current_tallies` and the site adds a new record in `past_tallies` for that user/round in every daily snapshot going forward. Said another way: we keep daily records of every user that has touched a round forever, even if they never proofread another page.

At pgdp.net, 24 years of this has caught up with us. The `past_tallies` table is 50GB in size and contains 598,611,362 rows. Of those 594,322,594 are records where `tally_delta`=0 -- 99%.

# This PR
This PR converts `past_tallies` to a sparse table wherein we do not store entries where `tally_delta`==0. It includes an upgrade script to delete those records.

We use `past_tallies` in a few primary ways:
1. To get the sum of pages done over a specific time span eg: last week, last month
   * A sparse table format where we do not store rows where `tally_delta` == 0 isn't a problem for this one. We just `sum(tally_delta)` over a time range anyway -- we don't need the zeros.
2. To get the number of pages done at a specific time (usually "yesterday")
   * This one we could calculate from a sparse `past_tallies` pretty easily.
3. To get the number of pages done every day over some span for a table or a chart
   * Without some sort of secondary table that records all of the snapshot times we can't accurately generate the data because we don't know what days are in-between the records where `tally_delta` <> 0. `past_tallies` itself isn't set up to query against `timestamp` and there's nothing that guarantees that every snapshot will have at least one non-zero `tally_delta`. (This might be practically true on pgdp.net but isn't guaranteed to be true for all sites.)
4. In concert with `current_tallies` to determine how many pages they did between yesterday and today to add new entries into `past_tallies
   * We can programmatically determine this from a sparse `past_tallies` but we add a whole lot of computation for every snapshot in ways that just don't make sense.

This PR introduces 1 new tables:
* `tally_snapshot_times` stores a timestamp of every snapshot. That's it. This one is useful for recreating charts and tables that might be filled with zeros.

And adds additional columns to another:
* `current_tallies` now stores the most recent snapshot as well. This allows us to take a snapshot very quickly and then populate `past_tallies`. It is also used as a very fast (index-based) lookup for "yesterday's" metrics which we do very often for rounds and users. And finally it is also a fast lookup for the most recent snapshot timestamp.

`past_tallies` is now touched in only the following places:
* `TallyBoard->take_snapshot()` -- update to only insert records where `tally_delta` != 0
* `TallyBoard->get_info_re_largest_delta()` -- unchanged, we still return info about the largest `tally_delta`. It already handles the case where the tally_holder does not exist in the table.
* `TallyBoard->get_deltas()` -- we initialize all possible snapshot times with zeros and fill in any non-zero entries.
* `TallyBoard->get_delta_sum()` -- unchanged
* `get_site_tally_grouped()` -- we initialize the snapshot times to zeros and then fill in the non-zero values from `past_tallies`.

# Testing
Testing/validating this is going to be a bit tricky because it involves changing data. I've created a separate database on TEST that is a copy of the current one. On this copy I've run the upgrade scripts and the sandbox below is using it. There is a crontab that refreshes the sparse DB after UTC midnight so they will only have at most a day of drift.

The sandbox has a nightly cronjob to take snapshots of the database but it is currently disabled. What this means is that the database will drift throughout a day and catch up at midnight UTC.

I suggest we test this in two phases:
1. Validate that the overall approach is sound, the non-snapshot-taking code works, and that the sparse DB sandbox matches the main TEST sandbox.  This is going to be confirming that the user's "yesterday" stats are correct and that the stats (graphs and tables) match across the rounds.
2. After we feel good with the code and the sparse format functionality I'll disable the sync and enable the snapshot crontab and we can validate that the nightly snapshot code is working like we want.

Sandbox: https://www.pgdp.org/~cpeel/c.branch/revamp-past-tallies-sandbox/

# Current `past_tallies` sizes on TEST
```
# Main DB
mysql> select count(*) from past_tallies;
+----------+
| count(*) |
+----------+
|  2304294 |
+----------+
1 row in set (1.69 sec)

# Sparse DB
mysql> select count(*) from past_tallies;
+----------+
| count(*) |
+----------+
|     9027 |
+----------+
1 row in set (0.01 sec)
```

cpeel commented:
I've created this testing code for a much more automated testing approach. The goal is to check that the data gathering functions in the current code against the main TEST database matches the new code against the sparse DB:
```php
<?php
$relPath = "./pinc/";
include_once($relPath."base.inc");

include($relPath."udb_user.php");
echo "DB: $db_name\n";

foreach (get_all_current_tallyboards() as $tallyboard) {
    if ($tallyboard->holder_type != 'S') {
        continue;
    }
    echo "tally_name: {$tallyboard->tally_name}, holder_type: {$tallyboard->holder_type}\n";
    $time_of_latest_snapshot = $tallyboard->get_time_of_latest_snapshot();
    echo "    get_time_of_latest_snapshot: $time_of_latest_snapshot\n";
    echo "    get_info_from_latest_snapshot:\n";
    dump_array($tallyboard->get_info_from_latest_snapshot(1));
    echo "    get_info_re_largest_delta:\n";
    dump_array($tallyboard->get_info_re_largest_delta(1));
    $timestamps = [
        [0, $time_of_latest_snapshot + 100],
        [1430438400, $time_of_latest_snapshot + 100],
        [1639958400, $time_of_latest_snapshot + 100],
        [1430438400, 1639958400],
    ];
    echo "    get_delta_sum:\n";
    foreach ($timestamps as [$start, $end]) {
        echo "        $start-$end: " . $tallyboard->get_delta_sum(1, $start, $end) . "\n";
    }
    echo "    get_deltas:\n";
    dump_array($tallyboard->get_deltas(1, 0));
}

function dump_array($array)
{
    echo preg_replace("/^/m", "        ", json_encode($array, JSON_PRETTY_PRINT | JSON_NUMERIC_CHECK));
    echo "\n";
}
```

I ran this against both and did a diff between them:
```
cpeel@ip-172-31-12-95:~/public_html/c.branch/revamp-past-tallies-sandbox$ php test-Tallyboard.php  > /tmp/sparse.txt
cpeel@ip-172-31-12-95:/data/htdocs/c$ php test-Tallyboard.php > /tmp/full.txt
cpeel@ip-172-31-12-95:/data/htdocs/c$ diff /tmp/full.txt /tmp/sparse.txt
1c1
< DB: dp_utf8test
---
> DB: dp_project_shrink
```

windymilla commented on SETUP/upgrade/21/20240414_populate_latest_shapshot_tallies.php:
```diff
@@ -0,0 +1,173 @@
+<?php
```
Typo in name of script? 
SETUP/upgrade/21/20240414_populate_latest_shapshot[** snapshot]_tallies.php

cpeel replied:
Good catch! I've renamed this script after merging `last_snapshot_tallies` into `current_tallies`.

windymilla reviewed (commented):
I couldn't find any data mismatches between the branch and the main code on TEST.

Apart from minor file name comment, LGTM!

jmdyck commented on pinc/TallyBoard.inc:
```diff
@@ -21,9 +21,19 @@
+// TallyBoards store and access data in 5 tables, 4 of which treat the
+// holder_type/holder_id/tally_name tuple as a primary key:
+// * current_tallies - stores the current tally, one per tuple
+// * past_tallies - stores historical tallies of deltas, one row per tuple
```
The primary key of `past_tallies` isn't `holder_type/holder_id/tally_name`, it also includes `timestamp`.

jmdyck commented on SETUP/db_schema.sql:
```diff
@@ -423,6 +437,15 @@ CREATE TABLE `special_days` (
+--
+
+CREATE TABLE `tally_snapshot_times` (
+  `timestamp` int(10) unsigned NOT NULL default '0',
```
This assumes that all tallyboards are snapshotted on the same schedule. The current db design doesn't require that, but `take_tally_snaphots.php` assumes it, so it's probably safe?

cpeel replied:
True. We could have it record a snapshot time for every Tallyboard but that seems excessive?

jmdyck replied:
> I want to keep each Tallyboard snapshot self-contained so I've kept the last snapshot timestamp as a column in `current_tallies`.

Hm.
- On the one hand, `current_tallies.last_snap_timestamp` suggests that, within a given TallyBoard, different holders could have different snapshot-times (which the TallyBoard code doesn't allow, and might cause some headaches if it ever did happen).
- On the other hand, the `tally_snapshot_times` table implies that all Tallyboards have the same snapshot times (which would violate the TallyBoard abstraction in the same way you object to above).

For an example of where the latter wouldn't work, consider a site that tweaked `take_tally_snapshots.php` to save some tallies nightly, and some weekly (but always at midnight). In the status quo, I think this would work. But with this PR, when that site calls `get_deltas` for a weekly tally (over some interval), it'll fetch all the daily snapshot times too, and set `$delta_at_` to 0 for all those other times. So the result will (incorrectly) look like a daily tally where (within each week) nothing happens for 6 days and then everything happens on one day.

Or consider a site where all TallyBoards are snapshotted daily, but some at noon and some at midnight. (This example seems less likely to me, but who knows.)

To resolve this, I suggest dropping `current_tallies.last_snap_timestamp`, and instead adding `tally_name` and `holder_type` columns to `tally_snapshot_times`, so each TallyBoard can have its own set of timestamps.

To get the latest timestamp for a given TallyBoard, we could just ask for `MAX(timestamp) WHERE ...`, which I'm guessing would be fairly quick, but could you also add an index on `(tally_name, holder_type, MAX(timestamp))`?

jmdyck replied:
Side note: I was looking at where we use time-of-latest-snapshot. `TallyBoard::get_info_from_latest_snapshot` fetches it and puts it into the array that it returns, but as far as I can tell, nobody uses that data. So that function could just not bother. [Not suggesting that for this PR.] On the other hand, maybe the recipients of that data *should* use the timestamp. (E.g., to say something like "as of $whenever server time".)

cpeel replied:
> To resolve this, I suggest dropping `current_tallies.last_snap_timestamp`, and instead adding `tally_name` and `holder_type` columns to `tally_snapshot_times`, so each TallyBoard can have its own set of timestamps.

That makes sense to me too, I'll work on that.

cpeel replied:
This is done. Note that it will take about 4 hours to populate `tally_snapshot_times` when we deploy this. Doing the `SELECT DISTINCT timestamp` for each tally_name/holder_type takes ~15 minutes because although it's an index scan, it takes that long just to (what I presume is) load the index from the disk to scan and I don't think the whole index will fit in memory.

srjfoo replied:
Bearing in mind my ignorance on this subject, does this imply that we should, at the very least, put out a general announcement to the site a day or two before the next code deployment that the site may/will be slow during "X" time frame? Or even have the site unavailable?

cpeel replied:
My suggestion on the deployment plan would be:

After the DB backup happens and the tallyboards are snapshotted, we start the data-population script. Its likely to slow down the site just like a backup does but probably for longer -- there should be no need to make the site unavailable. After the data-population script finishes we could do the code deployment that uses the new tables.

Sometime later (a day or so) we could start the script to prune tally_delta=0 values from `past_tallies` but that isn't required for the new code to work and can happen slowly over several days.

After that script finishes we'll want to do an optimize table on `past_tallies` to actually shrink the table. I have no idea how long that will take. According to the [MySQL docs](https://dev.mysql.com/doc/refman/8.0/en/optimize-table.html) it can be done entirely online with only needing a table lock at the beginning and at the end. That plus given `past_tallies` queries are now only in the stats code, we probably don't need a downtime window for that either but should start it after the daily tallyboard snapshot code runs.

jmdyck commented on pinc/TallyBoard.inc:
```diff
@@ -476,7 +524,7 @@ class TallyBoard
             "
             SELECT MAX(timestamp) as max_timestamp
-            FROM past_tallies
+            FROM latest_snapshot_tallies
```
If we have `tally_snapshot_times`, it would be simpler to get the time from there.

cpeel replied:
I thought about that, but this is more accurate since, as you've pointed out, taking snapshots isn't instantaneous.

jmdyck replied:
The `timestamp` that goes into `past_tallies` or `latest_snapshot_tallies` is `$ascribe_time`, which is set in `crontab/take_tally_snapshots.php`, and is constant for a given day's snapshot job. The fact that snapshots aren't instantaneous doesn't affect it.

Side comment: I'm not seeing where a new timestamp is added to `tally_snapshot_times` (outside of the upgrade scripts).

jmdyck reviewed (commented):
Not a full review, just some thoughts.

cpeel commented on SETUP/db_schema.sql:
```diff
@@ -118,6 +118,20 @@ CREATE TABLE `job_logs` (
+  `timestamp` int(10) unsigned NOT NULL default '0',
+  `tally_delta` int(8) NOT NULL default '0',
+  `tally_value` int(8) NOT NULL default '0',
+  PRIMARY KEY (`tally_name`,`holder_type`,`holder_id`)
```
I thought about that but I don't like the idea of making `current_tallies` not contain, well, just the current tallies. Then it becomes `current_and_last_snapshot_tallies` and we start muddling things together.

I suspect that taking snapshots is going to be faster since we're no longer relying on `past_tallies` to capture them at all -- we only use `current_tallies` and `last_snapshot_tallies`.

jmdyck replied:
> I thought about that but I don't like the idea of making `current_tallies` not contain, well, just the current tallies. Then it becomes `current_and_last_snapshot_tallies`

Personally, I think "current" is vague enough to include yesterday, but we could call it something else, like `recent_tallies`. 

> and we start muddling things together.

I think it's putting very closely related things together.

> I suspect that taking snapshots is going to be faster since we're no longer relying on `past_tallies` to capture them at all -- we only use `current_tallies` and `last_snapshot_tallies`.

I'm not sure it's going to be faster. It might even be slower, since we'd be replacing:
> for each holder on the tallyboard, insert into past_tallies
with:
> for each holder on the tallyboard, insert into latest_snapshot_tallies; then for each holder with a non-zero delta, insert into past_tallies

so it's strictly more insert commands, right? *Maybe* inserting into `latest_snapshot_tallies` will be faster than inserting into `past_tallies`, but I suspect the per-command overhead will dominate when we're doing it for thousands of users.

Not that that's an objection to this PR: having a sparse `past_tallies` is important, and a longer snapshot job would be an acceptable price to pay. *But* if we have the possibility of reducing the period of capture-uncertainty from 15-20 minutes down to (I don't know) a few seconds, I think we should look into it.

cpeel replied:
> I'm not sure it's going to be faster. It might even be slower, since we'd be replacing:
> 
> > for each holder on the tallyboard, insert into past_tallies
> 
> with:
> 
> > for each holder on the tallyboard, insert into latest_snapshot_tallies; then for each holder with a non-zero delta, insert into past_tallies
> 
> so it's strictly more insert commands, right?

You're eliding over the fact that we're no longer querying `past_tallies` to do the snapshot -- we're now query a primary key in a small table. The original query is non-negligible because while the query is indexed-based, the index is _huge_ because the table is huge and we have such broad indexes on it so its doing a lot of disk I/O just to load the index to do that query.

For most users we're not inserting into `latest_snapshot_tallies`, we're effectively updating a row based entirely on the value of the primary key -- so the DB doesn't have to allocate more space or manage indexes. We're inserting about ~1% of the current number of rows into `past_tallies` than we're doing now so we're likely doing 1% more update commands but 99% of them are more efficient.

I'll give the "throw everything into `current_tallies`" more thought but I suspect we're going to see pretty massive performance improvements even with the proposed design because we're taking `past_tallies` out of the picture as much as possible.

And note that structurally we wouldn't want to update all tallyboards in a single command as that would violate a couple of layers of abstraction -- we'd need to do it for a tallyboard at a time.

cpeel replied:
I've merged `last_snapshot_tallies` into `current_tallies`. I want to keep each Tallyboard snapshot self-contained so I've kept the last snapshot timestamp as a column in `current_tallies`.

cpeel replied:
Something to note about doing the update with an `UPDATE` over many rows is that it will lock those rows for as long as it takes to do the update. Presumably that is on the order of a second or two -- the table is simple and only has 100k rows -- but anything that tries to update during that window (when someone proofreads a page) that will block until the update completes.

cpeel replied:
I copied the `current_tallies` table from PROD to a new database on TEST and ran the UPDATE over tally_name=P1 and holder_type=U (the largest single Tallyboard) and it took <3 seconds:
```
Query OK, 67500 rows affected (2.33 sec)
Rows matched: 67827  Changed: 67500  Warnings: 0
```

TEST is slower than PROD so we should be good to go here.

jmdyck replied:
> You're eliding over the fact that we're no longer querying `past_tallies` to do the snapshot

Yup, I left that out because I thought it was negligible because of the index, but apparently not. (At least, not for the (P1,U) tallyboard, for the others it's pretty quick). (Once again, MySQL defies my intuitions.)

> And note that structurally we wouldn't want to update all tallyboards in a single command as that would violate a couple of layers of abstraction -- we'd need to do it for a tallyboard at a time.

True, it would disrepect the TallyBoard abstraction. (Not sure what other layer you're thinking of.)

So I was going to suggest a per-TallyBoard version of the 'capture everyone at once' command, but it looks like you've already got that.


> Something to note about doing the update with an `UPDATE` over many rows is that it will lock those rows for as long as it takes to do the update. Presumably that is on the order of a second or two -- the table is simple and only has 100k rows -- but anything that tries to update during that window (when someone proofreads a page) that will block until the update completes.

Right, so if someone saves-as-done (or reopens) a page in that small window, the site's response will be a bit slower than usual. But I'm guessing people won't find it too anomalous.


> I copied the `current_tallies` table from PROD to a new database on TEST and ran the UPDATE over tally_name=P1 and holder_type=U (the largest single Tallyboard) and it took <3 seconds:
> 
> TEST is slower than PROD so we should be good to go here.

Yay!

cpeel commented:
I've rebased the branch to fix the merge conflict.

I've updated the cronjob on TEST so that just before every UTC midnight the sandbox DB gets updated from the main TEST DB and the migration scripts run on it. Then just past midnight the nightly snapshot happens for the snapshot DB.

jmdyck commented on pinc/page_tally.inc:
```diff
@@ -414,31 +409,67 @@ function get_site_tally_grouped(
             )
         WHERE
-            past_tallies.tally_name = '%s'
+            $where_time
```
On the other hand, maybe this query should be rewritten to not involve `tally_snapshot_times`. (I.e., just fetch data from `site_tally_goals` with the appropriate constraining, grouping, and summing.)

E.g., say that snapshotting was changed to an hourly cadence. I think that, for a given day, this query would join the 24 `tally_snapshot_times` rows with 24 copies of the goal for that day, causing `sum(goal)` to be 24 times larger than it should be. (And if it were changed to a weekly cadence, that would only pick up 1/7th of the goal entries.)

Of course, the (potential) mismatch between per-date goals and arbitrary-cadence snapshots is a pre-existing problem (and it's not entirely clear what this script should do in all cases). It's maybe a bit more obvious (and avoidable) now that we're not trying to get all the data in a single query.

cpeel replied:
I removed the join altogether as it's really not needed, to your point. Now we're initializing the dataset with the goals and then overlaying the tally deltas.

jmdyck reviewed (changes requested):
I kind of lost steam in `get_site_tally_grouped`, as my brain is tired.

Also, I only looked at places that were different from master. A thorough review would also look for places that *should* change but haven't.

cpeel commented:
These are great feedback Michael, thanks. I'll work through them tonight.

cpeel commented on SETUP/upgrade/21/20240414_update_current_tallies.php:
```diff
@@ -0,0 +1,148 @@
+
+echo "\nDone!\n";
+
+// We have to use our own function in the upgrade script rather than the one
+// in Tallyboard because it will already be trying to use the table we're
+// trying to populate!
```
> But it doesn't apply to the pgdp deployment plan you outline https://github.com/DistributedProofreaders/dproofreaders/pull/1199#discussion_r1600383854, because we'll be running the upgrade scripts before the new `TallyBoard.inc` is installed, right?

Correct. But I didn't want to assume that's how other sites might be doing it.

> On the other hand, what if (for each tallyboard) we populate tally_snapshot_times before the new columns in current_tallies? Then we could use Tallyboard::get_time_of_latest_snapshot, and drop this extra function. I think I like that better.

I see what you mean, but I like being explicit about what's going on in the upgrade script. Its possible that we, or others, might do something out of order / run something manually / need to rerun pieces out of order / etc and having a function that always works vs works based on half of the upgrade script seems safer overall.

I'll update the comment above it though, as your version is clearer.

jmdyck replied:
> Its possible that we, or others, might do something out of order / run something manually / need to rerun pieces out of order / etc and having a function that always works vs works based on half of the upgrade script seems safer overall.

Okay in general, but could you give a specific example of something that we or others might do that would cause my suggested change to malfunction (that wouldn't also cause the existing PR code to malfunction)?

---

Also, if we're recognizing the possibility that someone might need to rerun this script, then it seems like the `ALTER TABLE` and `CREATE TABLE` code at the start should allow for that. Currently, the script will die with an error if you re-execute either of those, won't it?

---

Are we allowing for the possibility that a snapshot might be running at the same time as this script? I'm pretty sure we aren't. We should maybe have a comment at the top saying something like:
`past_tallies` must remain unchanged over the run (+ possible reruns) of this script, so choose a time well away from any snapshot (or any ad hoc query that could change `past_tallies`).

On second thought, we should add more stuff to `SETUP/UPGRADE.md` (not necessarily this PR):
- Disable general access to your site. (Maybe put it in maintenance mode.)
- Disable cron jobs.
- Generally, it's important that, when running upgrade scripts, **nothing else is changing your database**. [There are exceptions, but we don't want to get into them.]

cpeel replied:
> > Its possible that we, or others, might do something out of order / run something manually / need to rerun pieces out of order / etc and having a function that always works vs works based on half of the upgrade script seems safer overall.
> 
> Okay in general, but could you give a specific example of something that we or others might do that would cause my suggested change to malfunction (that wouldn't also cause the existing PR code to malfunction)?

Another site might decide to, for whatever reason, run the contents of the script out of order due to some special something they've done in their DB. Or maybe they had some silent failure in the first half of the script and then the second half runs.

Both methods work. Keeping the additional function seems to me more resilient (and it's already written and works so why change it?).

> Also, if we're recognizing the possibility that someone might need to rerun this script, then it seems like the `ALTER TABLE` and `CREATE TABLE` code at the start should allow for that. Currently, the script will die with an error if you re-execute either of those, won't it?

I'm again thinking of the case where there is some failure somewhere and the admin has to hack the script to pick up where it left off. In that case the user can just comment out the `mysqli_query()` for those and move on.

> Are we allowing for the possibility that a snapshot might be running at the same time as this script? I'm pretty sure we aren't. We should maybe have a comment at the top saying something like: `past_tallies` must remain unchanged over the run (+ possible reruns) of this script, so choose a time well away from any snapshot (or any ad hoc query that could change `past_tallies`).
> 
> On second thought, we should add more stuff to `SETUP/UPGRADE.md` (not necessarily this PR):
> 
>     * Disable general access to your site. (Maybe put it in maintenance mode.)
> 
>     * Disable cron jobs.
> 
>     * Generally, it's important that, when running upgrade scripts, **nothing else is changing your database**. [There are exceptions, but we don't want to get into them.]

Yes, agreed on both the comment in the script and the update to the `UPGRADE.md`.

jmdyck replied:
> > [...] could you give a specific example of something that we or others might do that would cause my suggested change to malfunction (that wouldn't also cause the existing PR code to malfunction)?
> 
> Another site might decide to, for whatever reason, run the contents of the script out of order due to some special something they've done in their DB. Or maybe they had some silent failure in the first half of the script and then the second half runs.

To make sure I understand, you're saying if we followed my suggestion (change the script to populate `tally_snapshot_times` before the new columns in `current_tallies` and call `Tallyboard::get_time_of_latest_snapshot` in the latter), and then a site decided to do those parts in the opposite order, and they had already installed the new `Tallyboard.inc`, then `Tallyboard::get_time_of_latest_snapshot` wouldn't work.

Okay, that's true, so I won't push on that suggestion any more.

But it seems like there's no end to things that other sites *might* do. Is there a line beyond which we don't bother to accommodate? Maybe not a well-defined line, but some rough rule of thumb?

> > ... Currently, the script will die with an error if you re-execute either of those, won't it?

> I'm again thinking of the case where there is some failure somewhere and the admin has to hack the script to pick up where it left off.

Ah, what I thought you meant was: the script should be written so that it can be re-run without hacking the script. (It doesn't have to be able to 'pick up where it left off', but it should be able to speed through the stuff that it's already done without error.) I thought that's how some past upgrade scripts had been written.

cpeel replied:
> But it seems like there's no end to things that other sites _might_ do. Is there a line beyond which we don't bother to accommodate? Maybe not a well-defined line, but some rough rule of thumb?

I'm sure there is but I don't know where it is honestly. It would be so much easier to quantify if we knew of a site that was using our code and taking our updates. In the spirit of open source, and because it forces us into some good habits, I want to keep releasing code that others can consume but man does it feel like we're jumping through a lot of hoops for no apparent need.

My minimum bar is "the upgrade scripts should run happy path for users assuming they took the site offline first" but also knowing that we do a lot of special things for pgdp.net that bypass the upgrade scripts`*` may mean that others do too. Indeed its possible that we do some special hacks for pgdp.net in this very same script if the default version takes too long or is too resource intensive. Such as populating the timestamps for S,P1 and then just doing  select insert for (S,P2), (S,P3), etc.

`*` like not converting `page_events` to UTF-8 IIRC.

> Ah, what I thought you meant was: the script should be written so that it can be re-run without hacking the script. (It doesn't have to be able to 'pick up where it left off', but it should be able to speed through the stuff that it's already done without error.) I thought that's how some past upgrade scripts had been written.

We have done that in the past. This one is more challenging because we would have to detect that the additional columns were added to `current_tallies` and skip that update. The `CREATE TABLE` could be done with the `IF NOT EXISTS` but there's no workaround for the alter that I'm aware of.

srjfoo replied:
> My minimum bar is "the upgrade scripts should run happy path for users assuming they took the site offline first" ...

Is it fair to state that outright in the interest of simplifying things? (Or would it?) I know it's going to be a complex upgrade, so, hopefully anyone using our code should, also.

cpeel replied:
Every variation that Michael and I have talked about will work for the happy path. At this point we're mostly bikeshedding 😁

cpeel commented:
Latest commit addresses the good feedback from @jmdyck.

The aforementioned validation script still shows no functional differences between the functions in `master` and those in this sandbox.

cpeel commented:
My intent is to get this merged in before June 8th and rolled out in the June 12th deployment.
